Documentation for DeviceRedundancyGroup model (#2697)
nautobot/nautobot: nautobot/docs/models/dcim/deviceredundancygroups.md

> # Device Redundancy Groups

Reviewer: ```suggestion
# Device Redundancy Groups

+++ 1.5.0

```

> Device Redundancy Groups are used to represent logical relationships between different devices. Typically, a redundancy group could represent a failover pair, failover group, or a load sharing cluster.

Reviewer: I like the usage examples, but I would have expected a bit more information here about the model itself and how to set it up in the first, especially with regard to fields like `device_redundancy_group_priority` in the mix.

```python
import pynautobot
import json

query = """
query ($device_name: [String]) {
    devices(name__ie: $device_name) {
        name
        device_redundancy_group {
            name
            members {
                name
                device_redundancy_group_priority
                interfaces(name__ie: "failover-link") {
                    type
                    name
                    ip_addresses {
                        host
                        prefix_length
                    }
                    parent_interface {
                        name
                        type
                    }
                }
            }
        }
    }
}
"""

def get_gql_failover_details(device_name):
    variables = {"device_name": device_name}
    nb = pynautobot.api(
        url="http://localhost:8080",
        token="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    )
    return nb.graphql.query(query=query, variables=variables)
```

### Retrieving the data - Primary Failover Unit ("nyc-fw-primary")

An example data returned from Nautobot is presented below.

```python
>>> primary_unit_details = get_gql_failover_details("nyc-fw-primary").json
```

```json

> {'data': {'devices': [{'name': 'nyc-fw-primary', 'device_redundancy_group': {'name': 'nyc-firewalls', 'members': [{'name': 'nyc-fw-primary', 'device_redundancy_group_priority': 100, 'interfaces': [{'type': 'VIRTUAL', 'name': 'failover-link', 'ip_addresses': [{'host': '172.27.48.0', 'prefix_length': 31}], 'parent_interface': {'name': 'gigabitethernet0/3', 'type': 'A_1000BASE_T'}}]}, {'name': 'nyc-fw-secondary', 'device_redundancy_group_priority': 50, 'interfaces': [{'type': 'VIRTUAL', 'name': 'failover-link', 'ip_addresses': [{'host': '172.27.48.1', 'prefix_length': 31}], 'parent_interface': {'name': 'gigabitethernet0/3', 'type': 'A_1000BASE_T'}}]}]}}]}}

Reviewer: Can we use `json.dumps` or similar here and below so that the JSON is more human-readable?

```

### Retrieving the data - Secondary Failover Unit ("nyc-fw-secondary")

An example data returned from Nautobot is presented below.

```python
>>> secondary_unit_details = get_gql_failover_details("nyc-fw-secondary").json
```

```json
{'data': {'devices': [{'name': 'nyc-fw-secondary', 'device_redundancy_group': {'name': 'nyc-firewalls', 'members': [{'name': 'nyc-fw-primary', 'device_redundancy_group_priority': 100, 'interfaces': [{'type': 'VIRTUAL', 'name': 'failover-link', 'ip_addresses': [{'host': '172.27.48.0', 'prefix_length': 31}], 'parent_interface': {'name': 'gigabitethernet0/3', 'type': 'A_1000BASE_T'}}]}, {'name': 'nyc-fw-secondary', 'device_redundancy_group_priority': 50, 'interfaces': [{'type': 'VIRTUAL', 'name': 'failover-link', 'ip_addresses': [{'host': '172.27.48.1', 'prefix_length': 31}], 'parent_interface': {'name': 'gigabitethernet0/3', 'type': 'A_1000BASE_T'}}]}]}}]}}
```

### Processing the data - Primary Failover Unit ("nyc-fw-primary")

Before rendering the data, we will process the data to simplify data variables within template:

```python

> >>> hostname = primary_unit_details['data']['devices'][0]['name']

Reviewer: we already knew `hostname` was `nyc-fw-primary` as it was an input to the query, we shouldn't need to re-derive it?

>>> hostname
'nyc-fw-primary'
>>>
>>> failover_members = primary_unit_details['data']['devices'][0]['device_redundancy_group']['members']
>>>
>>> device = [failover_member for failover_member in failover_members if failover_member['name'] == hostname][0]
>>>
>>> device
{'name': 'nyc-fw-primary', 'device_redundancy_group_priority': 100, 'interfaces': [{'type': 'VIRTUAL', 'name': 'failover-link', 'ip_addresses': [{'address': '172.27.48.0/31'}], 'parent_interface': {'name': 'gigabitethernet0/3', 'type': 'A_1000BASE_T'}}]}
>>>
>>> failover_peer = [failover_member for failover_member in failover_members if failover_member['name'] != hostname][0]
>>>
>>> device_failover_interface = device['interfaces'][0]
>>> peer_failover_interface = failover_peer['interfaces'][0]
```

### Creating Cisco ASA Configuration Template - Common for Primary and Secondary Units

> Following snippet represents an example Juniper BGP Configuration Template:

Reviewer: Which is it? Juniper BGP or Cisco ASA? :-)

Author: Thanks! Fixed - was having some git/Pycharm issues yesterday and committed un-completed version :-(

```python
# Configuration Template for Cisco ASA
template_body="""
!
hostname {{ device.name }}
!

> failover lan unit {{ priority_mapping.device_redundancy_group_priority }}

Reviewer: Where did `priority_mapping` come from?

Author: It is coming via `template.render()` kwargs, although added a few lines about it in the intro 👍

> failover lan interface {{ device.interfaces[0].name }} {{ device.interfaces[0].parent_interface.name }} 

Reviewer: Should this be:

```suggestion
failover lan interface {{ device_failover_interface.name }} {{ device_failover_interface.parent_interface.name }} 
```

!

> failover interface ip {{ device.interfaces[0].name }} {{ device_failover_interface.ip_addresses[0].host }}/{{ device_failover_interface.ip_addresses[0].prefix_length }} standby {{ peer_failover_interface.ip_addresses[0].host }}

Reviewer: ```suggestion
failover interface ip {{ device_failover_interface.name }} {{ device_failover_interface.ip_addresses[0].host }}/{{ device_failover_interface.ip_addresses[0].prefix_length }} standby {{ peer_failover_interface.ip_addresses[0].host }}
```

interface {{ device_failover_interface.parent_interface.name }} 
  no shutdown
!

> failover link {{ device.interfaces[0].name }} {{ device_failover_interface.parent_interface.name }} 

Reviewer: ```suggestion
failover link {{ device_failover_interface.name }} {{ device_failover_interface.parent_interface.name }} 
```

!
!failover ipsec pre-shared-key !Nautobot Secrets
!
failover
!
"""
```

> ### Rendering Juniper Jinja2 BGP Configuration Template with the data retrieved from GraphQL

Reviewer: Cisco ASA not Juniper BGP, presumably

Author: Thanks! Fixed - was having some git/Pycharm issues yesterday and committed un-completed version :-(

> Following snippet represents an example Juniper BGP Renderer Configuration:

Reviewer: ```suggestion
Following snippet represents an example Cisco ASA Configuration:
```

Author: Thanks! Fixed - was having some git/Pycharm issues yesterday and committed un-completed version :-(

!
!failover ipsec pre-shared-key !Nautobot Secrets
!
failover
!
```

## Example use of Device Redundancy Groups - Spine Redundancy in a Leaf and Spine (Clos) Topology 

Another example for the redundancy group use case could be a spine redundancy in the Leaf and Spine topology.
Spine redundancy is important while performing the Day-2 operations, such as OS-updates.

In this scenario, no more than 1 device participating in a Device Redundancy Group should be updated and rebooted at the same time. In order to track this, we will create a new Device custom field  named `upgrade_operational_state` and assign it one of the statues: `pre_upgrade`, `in_reboot`, `post_upgrade`. If a device with a spine role assigned is in state `in_reboot`, no other redundancy group members should be OS-upgraded at the same time.

### Querying for the data - Spine Redundancy in a Leaf and Spine (Clos) Topology 

To retrieve the data about devices forming a Spine redundancy group, we will use the following GraphQL query:

```text
query {
	device_redundancy_groups(name__ie: "nyc-spines"){
    name
    members {
      name
      device_role {
        slug
      }

>       _custom_field_data

Reviewer: Would be more elegant to use `cf_upgrade_operational_state` here.

> # Device Redundancy Groups

Reviewer: Should probably add a include of this file in `nautobot/docs/core-functionality/devices.md` for discoverability.

> # Device Redundancy Groups

Reviewer 2: Sorry to jump in at the last minute but shouldn't this file be named singular, not plural, e.g. `deviceredundancygroup.md`?

Reviewer: Yes. Good catch!

Author: thx!

+++ 1.5.0

Device Redundancy Groups represent logical relationships between multiple devices. Typically, a redundancy group could represent a failover pair, failover group, or a load sharing cluster.
Device Redundancy Groups are created first, before the devices are assigned to the group.

A failover strategy represents intended operation mode of the group. Supported failover strategy are: Active/Active and Active/Standby.

> Secrets group are used to inform store secret information used by failover or a cluster of devices.

Reviewer: ```suggestion
Secrets groups could be used to inform store secret information used by failover or a cluster of devices.
```

> Device Redundancy Group Priority is a Device attribute defined during assigning a Device to a Device Redundancy Group. Device Redundancy Priority represents the priority the device has in the device redundancy group.

Reviewer: ```suggestion
Device Redundancy Group Priority is a Device attribute defined during assigning a Device to a Device Redundancy Group. This field represents the priority the device has in the device redundancy group.
```

## Example use of Device Redundancy Groups - Cisco ASA 5500 Series Active/Standby Failover

This document provides an example of generating a Cisco ASA device's desired failover configuration based on data stored in Nautobot Device Redundancy Groups. 
A GraphQL query is used to retrieve the relevant data, which is then rendered through a Jinja2 template to produce the desired configuration.

As one of the use cases for Device Redundacy Groups, introduced in Nautobot 1.5 release, is to model failover pairs, we will use the ASA 5500 Series in this example with details as follows:

* Firewall failover pair is composed of two devices named "nyc-fw-primary" and "nyc-fw-secondary"
* Each firewall device forming a failover cluster has a virtual interface dedicated for failover usage (named "failover-link") and addressed in 172.27.48.0/31 network
* Virtual failover interface has a physical parent interface assigned ("gigabitethernet0/3")
* Following redundancy group priorities are assigned in a failover pair:
  * Priority 100 for a Primary Failover unit
  * Priority 50 for a Secondary Failover unit

### Querying for the data

> To retrieve information about devices forming an ASA Failover pair, we will use GraphQL query and the `get_gql_failover_details` Python method.

Reviewer: ```suggestion
To retrieve information about devices forming an ASA Failover pair, we will use a GraphQL query and the `get_gql_failover_details` Python method.
```

                                }
                            ]
                        }
                    ]
                }
            }
        ]
    }
}
```

### Processing the data - Primary Failover Unit ("nyc-fw-primary")

Before rendering the data, we will process the data to simplify data variables within the template:

```python
>>> failover_members = primary_unit_details['data']['devices'][0]['device_redundancy_group']['members']
>>>
>>> device = [failover_member for failover_member in failover_members if failover_member['name'] == hostname][0]
>>>
>>> device
{'name': 'nyc-fw-primary', 'device_redundancy_group_priority': 100, 'interfaces': [{'type': 'VIRTUAL', 'name': 'failover-link', 'ip_addresses': [{'address': '172.27.48.0/31'}], 'parent_interface': {'name': 'gigabitethernet0/3', 'type': 'A_1000BASE_T'}}]}
>>>
>>> failover_peer = [failover_member for failover_member in failover_members if failover_member['name'] != hostname][0]
>>>
>>> device_failover_interface = device['interfaces'][0]
>>> peer_failover_interface = failover_peer['interfaces'][0]
```

### Creating Cisco ASA Configuration Template - Common for Primary and Secondary Units

> Following snippet represents an example Cisco ASA Failover configuration template:

Reviewer: ```suggestion
The following snippet represents an example Cisco ASA failover configuration template:
```
